pkg/cloudsql/adapter.go

```diff
@@ -0,0 +1,263 @@
+		PGDriver:       pgPool,
+		CloudSQLConfig: config,
+		CloudMonitoringClient: &CloudMonitoringClient{
+			client: client,
+			ctx:    ctx,
+		},
+		CloudSQLAdminClient: &SqlAdminClient{
+			client: sqladminService,
+		},
+		GuardrailSettings: &guardrailSettings,
+	}
+
+	c.InitCollectors(c.Collectors())
+
+	return c, nil
+}
+
+func (adapter *CloudSQLAdapter) ApplyConfig(proposedConfig *agent.ProposedConfigResponse) error {
+	// TODO: need to make sure that the last update has been applied - I think this is fairly "edge casey"
+	// but the update is non-zero in time so we should gracefully handle
+	// also the knobs that need restarts!
+
+	adapter.Logger().Infof("Applying config")
+	adapter.Logger().Infof("Proposed Config: %v", proposedConfig)
+	adapter.Logger().Infof("Knobs to Override: %v", proposedConfig.KnobsOverrides)
+
+	flags := []*sqladmin.DatabaseFlags{}
+
+	for _, knob := range proposedConfig.KnobsOverrides {
+		knobConfig, err := parameters.FindRecommendedKnob(proposedConfig.Config, knob)
+		if err != nil {
+			return fmt.Errorf("failed to find recommended knob: %v", err)
+		}
+		fmtValue, err := knobConfig.GetSettingValue()
+		if err != nil {
+			return fmt.Errorf("failed to get setting value: %v", err)
+		}
+
+		param := &sqladmin.DatabaseFlags{
+			Name:  knobConfig.Name,
+			Value: fmtValue,
+		}
+
+		flags = append(flags, param)
+	}
+
+	adapter.CloudSQLAdminClient.applyFlags(adapter.CloudSQLConfig.ProjectID, adapter.CloudSQLConfig.DatabaseName, flags)
+
+	return nil
+}
+
+func (adapter *CloudSQLAdapter) GetActiveConfig() (agent.ConfigArraySchema, error) {
+	adapter.Logger().Debugf("Getting Active Config")
+
+	return pg.GetActiveConfig(adapter.PGDriver, context.Background(), adapter.Logger())
+}
+
+func (adapter *CloudSQLAdapter) GetSystemInfo() ([]metrics.FlatValue, error) {
+	adapter.Logger().Debugf("Getting System Info")
+
+	// Get PostgreSQL version and max connections from database
+	pgVersion, err := pg.PGVersion(adapter.PGDriver)
+	if err != nil {
+		return nil, err
+	}
+
+	maxConnections, err := pg.MaxConnections(adapter.PGDriver)
+	if err != nil {
+		return nil, err
+	}
+
+	version, err := metrics.PGVersion.AsFlatValue(pgVersion)
+	if err != nil {
+		adapter.Logger().Errorf("Error creating PostgreSQL version metric: %v", err)
+		return nil, err
+	}
+
+	maxConnectionsMetric, err := metrics.PGMaxConnections.AsFlatValue(maxConnections)
+	if err != nil {
+		adapter.Logger().Errorf("Error creating max connections metric: %v", err)
+		return nil, err
+	}
+
+	cpuCount, err := GetCPUCount(adapter.CloudMonitoringClient, adapter.CloudSQLConfig.ProjectID, adapter.CloudSQLConfig.DatabaseName)
+	if err != nil {
+		adapter.Logger().Errorf("Error fetching CPU count: %v", err)
+		return nil, err
+	}
+
+	cpuCountMetric, err := metrics.NodeCPUCount.AsFlatValue(int64(cpuCount.GetDoubleValue()))
+	if err != nil {
+		adapter.Logger().Errorf("Error creating NodeCPUCount metric: %v", err)
+		return nil, err
+	}
```

Comment:
Add a TODO/Issue for this given they can have fractional count. I'm worried about it rounding down to `0` and causing issues. Not sure if always rounding up makes sense, but it might make sense to do:

```
1 if cpuFractionalCount < 1 else int(cpuFractionalCount
```